File: packages/decision-engine/tradeoffs/sensitivity.ts
/**
 * Weight Sensitivity and Stability Analysis Engine.
 */

import { DecisionCandidate, DecisionCriterion } from "../criteria/criterion-types";
import { DeterministicScorer } from "../evaluation/score";

export type DecisionStability = "ROBUST" | "MODERATELY_STABLE" | "SENSITIVE" | "UNSTABLE";

export interface SensitivityAnalysisResult {
  criterionId: string;
  originalWeight: number;
  testedWeight: number;
  originalWinner: string;
  newWinner: string;
  isRankingChanged: boolean;
}

export class SensitivityAnalyzer {
  public static analyze(
    candidates: DecisionCandidate[],
    criteria: DecisionCriterion[],
    rawMatrix: Record<string, Record<string, number>>
  ): { stability: DecisionStability; analyses: SensitivityAnalysisResult[] } {
    if (candidates.length < 2) {
      return { stability: "ROBUST", analyses: [] };
    }

    // Determine base winner
    const baseScores = candidates.map((c) => ({
      name: c.name,
      score: DeterministicScorer.calculateCandidateScore(c, criteria, rawMatrix[c.candidateId] || {}).totalScore,
    }));
    baseScores.sort((a, b) => b.score - a.score);
    const baseWinner = baseScores[0].name;

    const analyses: SensitivityAnalysisResult[] = [];
    let rankChanges = 0;

    for (const crit of criteria) {
      // Perturb weight by +50% and -50%
      for (const factor of [0.5, 1.5]) {
        const perturbed = criteria.map((c) =>
          c.criterionId === crit.criterionId ? { ...c, weight: c.weight * factor } : c
        );

        const perturbedScores = candidates.map((c) => ({
          name: c.name,
          score: DeterministicScorer.calculateCandidateScore(c, perturbed, rawMatrix[c.candidateId] || {}).totalScore,
        }));
        perturbedScores.sort((a, b) => b.score - a.score);
        const newWinner = perturbedScores[0].name;

        if (newWinner !== baseWinner) {
          rankChanges++;
          analyses.push({
            criterionId: crit.criterionId,
            originalWeight: crit.weight,
            testedWeight: Number((crit.weight * factor).toFixed(2)),
            originalWinner: baseWinner,
            newWinner,
            isRankingChanged: true,
          });
        }
      }
    }

    let stability: DecisionStability = "ROBUST";
    if (rankChanges > 2) {
      stability = "UNSTABLE";
    } else if (rankChanges > 0) {
      stability = "SENSITIVE";
    }

    return { stability, analyses };
  }
}

File: backend/workline/decision/sensitivity.py
"""Sensitivity Analysis for Decision Criteria Weight Perturbations."""

from typing import Dict, List, Tuple
from backend.workline.decision.models import (
    DecisionCandidate,
    DecisionCriterion,
    SensitivityAnalysis,
)
from backend.workline.decision.scoring import DeterministicScorer


class SensitivityAnalyzer:
    """Evaluates stability of the top recommendation against weight variations."""

    @classmethod
    def analyze(
        cls,
        candidates: List[DecisionCandidate],
        criteria: List[DecisionCriterion],
        raw_matrix: Dict[str, Dict[str, float]],
    ) -> Tuple[str, List[SensitivityAnalysis]]:
        if len(candidates) < 2:
            return "ROBUST", []

        # Find base winner
        scored_base = []
        for c in candidates:
            score, _ = DeterministicScorer.calculate_score(c, criteria, raw_matrix.get(c.candidate_id, {}))
            scored_base.append((c.name, score))
        scored_base.sort(key=lambda x: x[1], reverse=True)
        base_winner = scored_base[0][0]

        analyses: List[SensitivityAnalysis] = []
        rank_changes = 0

        for crit in criteria:
            for factor in [0.5, 1.5, 2.0]:
                perturbed_criteria = [
                    crit.model_copy(update={"weight": crit.weight * factor})
                    if c.criterion_id == crit.criterion_id
                    else c
                    for c in criteria
                ]

                scored_perturbed = []
                for c in candidates:
                    score, _ = DeterministicScorer.calculate_score(c, perturbed_criteria, raw_matrix.get(c.candidate_id, {}))
                    scored_perturbed.append((c.name, score))
                scored_perturbed.sort(key=lambda x: x[1], reverse=True)
                new_winner = scored_perturbed[0][0]

                if new_winner != base_winner:
                    rank_changes += 1
                    analyses.append(
                        SensitivityAnalysis(
                            criterion_id=crit.criterion_id,
                            original_weight=crit.weight,
                            tested_weight=round(crit.weight * factor, 2),
                            original_winner=base_winner,
                            new_winner=new_winner,
                            is_ranking_changed=True,
                        )
                    )

        stability = "ROBUST"
        if rank_changes >= 3:
            stability = "UNSTABLE"
        elif rank_changes > 0:
            stability = "SENSITIVE"

        return stability, analyses
